render/gles2/texture.c

```diff
@@ -318,7 +360,48 @@ struct wlr_texture *gles2_texture_from_dmabuf(struct wlr_renderer *wlr_renderer,
 
 	wlr_egl_restore_context(&prev_ctx);
 
-	wl_list_insert(&renderer->textures, &texture->link);
+	return &texture->wlr_texture;
+}
+
+static void texture_handle_buffer_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_gles2_texture *texture =
+		wl_container_of(listener, texture, buffer_destroy);
+	gles2_texture_destroy(texture);
+}
+
+struct wlr_texture *gles2_texture_from_buffer(struct wlr_renderer *wlr_renderer,
+		struct wlr_buffer *buffer) {
+	struct wlr_gles2_renderer *renderer = gles2_get_renderer(wlr_renderer);
+
+	struct wlr_dmabuf_attributes dmabuf;
+	if (!wlr_buffer_get_dmabuf(buffer, &dmabuf)) {
+		return false;
+	}
+
+	struct wlr_gles2_texture *texture;
+	wl_list_for_each(texture, &renderer->textures, link) {
```

Comment:
That's a recurring theme in wlroots, and I've been thinking for a while that something like a hashmap would allow us to optimize more cases. I've written down a proposal here: https://github.com/swaywm/wlroots/issues/2927